hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/FsDatasetImpl.java

```diff
@@ -578,7 +578,13 @@ public void removeVolumes(
           // Unlike updating the volumeMap in addVolume(), this operation does
           // not scan disks.
           for (String bpid : volumeMap.getBlockPoolList()) {
-            List<ReplicaInfo> blocks = new ArrayList<>();
+            List<ReplicaInfo> blocks;
+            if (blkToInvalidate.containsKey(bpid)) {
+              blocks = blkToInvalidate.get(bpid);
+            } else {
+              blocks = new ArrayList<>();
+              blkToInvalidate.put(bpid, blocks);
+            }
```

Comment:
Just thinking - this fix will need to go on all branches back into the 2.8 branch, which I think still uses Java 7. Therefore it would be ideal to have the computeIfAbsent method for trunk and 3.x but we will need a patch like this existing one for branch 2.

Reply:
@sodonnel Thank you for reviewing this!

Okay, I will change the code for your review here. And after this is committed to trunk, I will make a new PR for branch 2, which is like the original code.